File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/routeDetails/RouteDetailsView.kt
package com.mbta.tid.mbta_app.android.routeDetails

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.ModalRoutes
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.component.StarIcon
import com.mbta.tid.mbta_app.android.state.getGlobalData
import com.mbta.tid.mbta_app.android.util.fromHex
import com.mbta.tid.mbta_app.model.routeDetailsPage.RouteDetailsContext
import com.mbta.tid.mbta_app.viewModel.IErrorBannerViewModel

@Composable
fun RouteDetailsView(
    selectionId: String,
    context: RouteDetailsContext,
    onOpenStopDetails: (String) -> Unit,
    onBack: () -> Unit,
    onClose: () -> Unit,
    openModal: (ModalRoutes) -> Unit,
    errorBannerViewModel: IErrorBannerViewModel,
) {
    val globalData = getGlobalData("RouteDetailsView")
    val lineOrRoute = globalData?.getLineOrRoute(selectionId)
    if (lineOrRoute == null) {
        LoadingRouteStopListView(context, errorBannerViewModel)
        return
    }

    RouteStopListView(
        lineOrRoute,
        context,
        globalData,
        onClick = {
            when (it) {
                is RouteDetailsRowContext.Details -> onOpenStopDetails(it.stop.id)
                is RouteDetailsRowContext.Favorites -> it.onTapStar()
            }
        },
        onClickLabel = {
            when (it) {
                is RouteDetailsRowContext.Details -> null
                is RouteDetailsRowContext.Favorites ->
                    if (it.isFavorited) {
                        stringResource(R.string.remove_favorite)
                    } else {
                        stringResource(R.string.add_favorite_label)
                    }
            }
        },
        onBack = onBack,
        onClose = onClose,
        openModal = openModal,
        errorBannerViewModel,
        defaultSelectedRouteId = selectionId.takeUnless { it == lineOrRoute.id },
        rightSideContent = { rowContext, modifier ->
            when (rowContext) {
                is RouteDetailsRowContext.Details ->
                    Image(
                        painterResource(id = R.drawable.baseline_chevron_right_24),
                        contentDescription = null,
                        modifier = modifier.width(8.dp),
                        colorFilter = ColorFilter.tint(colorResource(R.color.deemphasized)),
                    )
                is RouteDetailsRowContext.Favorites -> {
                    val favoriteCD = stringResource(R.string.favorite_stop)
                    StarIcon(
                        rowContext.isFavorited,
                        Color.fromHex(lineOrRoute.backgroundColor),
                        contentDescription = if (rowContext.isFavorited) favoriteCD else "",
                    )
                }
            }
        },
    )
}

File: iosApp/iosApp/Pages/RouteDetails/RouteDetailsView.swift
//
//  RouteDetailsView.swift
//  iosApp
//
//  Created by Melody Horn on 7/7/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct RouteDetailsView: View {
    let selectionId: String
    let context: RouteDetailsContext
    let onOpenStopDetails: (String) -> Void
    let onBack: () -> Void
    let onClose: () -> Void
    let errorBannerVM: IErrorBannerViewModel

    @State var globalData: GlobalResponse?
    @State private var lineOrRoute: RouteCardData.LineOrRoute?
    let globalRepository: IGlobalRepository = RepositoryDI().global

    var body: some View {
        ScrollView([]) {
            if let globalData, let lineOrRoute {
                RouteStopListView(
                    lineOrRoute: lineOrRoute,
                    context: context,
                    globalData: globalData,
                    onClick: { row in
                        switch row {
                        case let .details(stop: stop): onOpenStopDetails(stop.id)
                        case let .favorites(isFavorited: _, onTapStar: onTapStar): onTapStar()
                        }
                    },
                    onBack: onBack,
                    onClose: onClose,
                    errorBannerVM: errorBannerVM,
                    defaultSelectedRouteId: selectionId == lineOrRoute.id ? nil : selectionId,
                    rightSideContent: rightSideContent
                )
            } else {
                loadingBody()
            }
        }
        .onAppear {
            getGlobal()
        }
        .onChange(of: globalData) { globalData in
            lineOrRoute = globalData?.getLineOrRoute(lineOrRouteId: selectionId)
        }
    }

    @MainActor
    func activateGlobalListener() async {
        for await globalData in globalRepository.state {
            self.globalData = globalData
        }
    }

    func getGlobal() {
        Task(priority: .high) {
            await activateGlobalListener()
        }
        Task {
            await fetchApi(
                errorKey: "NearbyTransitView.getGlobal",
                getData: { try await globalRepository.getGlobalData() },
                onRefreshAfterError: { @MainActor in getGlobal() }
            )
        }
    }

    @ViewBuilder private func loadingBody() -> some View {
        let objects = ObjectCollectionBuilder()
        let mockRoute = RouteCardData.LineOrRouteRoute(route: objects.route { _ in })
        let mockGlobal = GlobalResponse(objects: objects)
        let toastVM = MockToastViewModel(initialState: ToastViewModel.StateHidden())
        RouteStopListContentView(
            lineOrRoute: mockRoute,
            parameters: .init(lineOrRoute: mockRoute, globalData: mockGlobal),
            selectedDirection: 0,
            setSelectedDirection: { _ in },
            selectedRouteId: mockRoute.id,
            setSelectedRouteId: { _ in },
            stopList: .none,
            context: context,
            globalData: globalData ?? GlobalResponse(objects: .init()),
            onClick: { _ in },
            onBack: {},
            onClose: {},
            errorBannerVM: errorBannerVM,
            rightSideContent: rightSideContent,
            toastVM: toastVM
        )
        .loadingPlaceholder()
    }

    @ViewBuilder private func rightSideContent(rowContext: RouteDetailsRowContext) -> some View {
        switch rowContext {
        case .details:
            Image(.faChevronRight)
                .resizable().frame(width: 8, height: 14).padding(.trailing, 8)
                .foregroundStyle(Color.deemphasized)
        case let .favorites(isFavorited: isFavorited, onTapStar: _):
            var starColor: Color {
                if let lineOrRoute {
                    Color(hex: lineOrRoute.backgroundColor)
                } else {
                    Color.text
                }
            }
            StarIcon(starred: isFavorited, color: starColor)
                .padding(.trailing, 8)
                .accessibilityLabel(isFavorited ? Text("favorite stop") : Text(verbatim: ""))
        }
    }
}
